File: core/src/com/bombbird/terminalcontrol/screens/settingsscreen/categories/DisplaySettingsScreen.java
package com.bombbird.terminalcontrol.screens.settingsscreen.categories;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.scenes.scene2d.Actor;
import com.badlogic.gdx.scenes.scene2d.ui.Image;
import com.badlogic.gdx.scenes.scene2d.ui.Label;
import com.badlogic.gdx.scenes.scene2d.ui.SelectBox;
import com.badlogic.gdx.scenes.scene2d.utils.ChangeListener;
import com.bombbird.terminalcontrol.TerminalControl;
import com.bombbird.terminalcontrol.screens.gamescreen.RadarScreen;
import com.bombbird.terminalcontrol.screens.settingsscreen.SettingsTab;
import com.bombbird.terminalcontrol.screens.settingsscreen.SettingsTemplateScreen;
import com.bombbird.terminalcontrol.utilities.saving.GameSaver;

public class DisplaySettingsScreen extends SettingsTemplateScreen {
    public SelectBox<String> trajectoryLine;
    public SelectBox<String> pastTrajectory;
    public SelectBox<String> mva;
    public SelectBox<String> ilsDash;
    public SelectBox<String> showUncontrolledTrail;
    public SelectBox<String> rangeCircle;
    public SelectBox<String> colour;

    public Label trajectoryLabel;
    public int trajectorySel;

    public Label pastTrajLabel;
    public int pastTrajTime;

    public Label mvaLabel;
    public boolean showMva;

    public Label ilsDashLabel;
    public boolean showIlsDash;

    public Label showUncontrolledLabel;
    public boolean showUncontrolled;

    public Label rangeCircleLabel;
    public int rangeCircleDist;

    public Label colourLabel;
    public int colourStyle;

    public DisplaySettingsScreen(TerminalControl game, RadarScreen radarScreen, Image background) {
        super(game, radarScreen, background);

        infoString = "Set the game display settings below.";
        loadUI(-1200, -200);

        setOptions();
    }

    /** Loads selectBox for display settings */
    @Override
    public void loadBoxes() {
        trajectoryLine = createStandardSelectBox();
        trajectoryLine.setItems("Off", "60 sec", "90 sec", "120 sec", "150 sec");
        trajectoryLine.addListener(new ChangeListener() {
            @Override
            public void changed(ChangeEvent event, Actor actor) {
                if ("Off".equals(trajectoryLine.getSelected())) {
                    trajectorySel = 0;
                } else {
                    trajectorySel = Integer.parseInt(trajectoryLine.getSelected().split(" ")[0]);
                }
            }
        });

        pastTrajectory = createStandardSelectBox();
        pastTrajectory.setItems("Off", "60 sec", "120 sec", "All");
        pastTrajectory.addListener(new ChangeListener() {
            @Override
            public void changed(ChangeEvent event, Actor actor) {
                if ("Off".equals(pastTrajectory.getSelected())) {
                    pastTrajTime = 0;
                } else if ("All".equals(pastTrajectory.getSelected())) {
                    pastTrajTime = -1;
                } else {
                    pastTrajTime = Integer.parseInt(pastTrajectory.getSelected().split(" ")[0]);
                }
            }
        });

        mva = createStandardSelectBox();
        mva.setItems("Show", "Hide");
        mva.addListener(new ChangeListener() {
            @Override
            public void changed(ChangeEvent event, Actor actor) {
                if ("Show".equals(mva.getSelected())) {
                    showMva = true;
                } else if ("Hide".equals(mva.getSelected())) {
                    showMva = false;
                } else {
                    Gdx.app.log(getClass().getName(), "Unknown MVA setting " + mva.getSelected());
                }
            }
        });

        ilsDash = createStandardSelectBox();
        ilsDash.setItems("Simple", "Realistic");
        ilsDash.addListener(new ChangeListener() {
            @Override
            public void changed(ChangeEvent event, Actor actor) {
                if ("Simple".equals(ilsDash.getSelected())) {
                    showIlsDash = false;
                } else if ("Realistic".equals(ilsDash.getSelected())) {
                    showIlsDash = true;
                } else {
                    Gdx.app.log(getClass().getName(), "Unknown ILS dash setting " + ilsDash.getSelected());
                }
            }
        });

        showUncontrolledTrail = createStandardSelectBox();
        showUncontrolledTrail.setItems("When selected", "Always");
        showUncontrolledTrail.addListener(new ChangeListener() {
            @Override
            public void changed(ChangeEvent event, Actor actor) {
                if ("Always".equals(showUncontrolledTrail.getSelected())) {
                    showUncontrolled = true;
                } else if ("When selected".equals(showUncontrolledTrail.getSelected())) {
                    showUncontrolled = false;
                } else {
                    Gdx.app.log(getClass().getName(), "Unknown uncontrolled trail setting " + ilsDash.getSelected());
                }
            }
        });

        rangeCircle = createStandardSelectBox();
        rangeCircle.setItems("Off", "5nm", "10nm", "15nm", "20nm");
        rangeCircle.addListener(new ChangeListener() {
            @Override
            public void changed(ChangeEvent event, Actor actor) {
                if ("Off".equals(rangeCircle.getSelected())) {
                    rangeCircleDist = 0;
                } else {
                    rangeCircleDist = Integer.parseInt(rangeCircle.getSelected().replace("nm", ""));
                }
            }
        });

        colour = createStandardSelectBox();
        colour.setItems("More colourful", "More standardised");
        colour.addListener(new ChangeListener() {
            @Override
            public void changed(ChangeEvent event, Actor actor) {
                if ("More colourful".equals(colour.getSelected())) {
                    colourStyle = 0;
                } else if ("More standardised".equals(colour.getSelected())) {
                    colourStyle = 1;
                } else {
                    Gdx.app.log(getClass().getName(), "Unknown colour style setting " + colour.getSelected());
                }
            }
        });
    }

    /** Loads labels for display settings */
    @Override
    public void loadLabel() {
        super.loadLabel();

        trajectoryLabel = new Label("Trajectory line: ", labelStyle);
        pastTrajLabel = new Label("Aircraft trail: ", labelStyle);
        mvaLabel = new Label("MVA altitude: ", labelStyle);
        ilsDashLabel = new Label("ILS display: ", labelStyle);
        showUncontrolledLabel = new Label("Show uncontrolled:\naircraft trail", labelStyle);
        rangeCircleLabel = new Label("Range rings:", labelStyle);
        colourLabel = new Label("Colour style:", labelStyle);
    }

    /** Loads actors for display settings into tabs */
    @Override
    public void loadTabs() {
        SettingsTab tab1 = new SettingsTab(this, 2);
        tab1.addActors(trajectoryLine, trajectoryLabel);
        tab1.addActors(pastTrajectory, pastTrajLabel);
        tab1.addActors(mva, mvaLabel);
        tab1.addActors(ilsDash, ilsDashLabel);
        tab1.addActors(showUncontrolledTrail, showUncontrolledLabel);
        tab1.addActors(rangeCircle, rangeCircleLabel);
        tab1.addActors(colour, colourLabel);

        settingsTabs.add(tab1);
    }

    /** Sets relevant options into select boxes */
    @Override
    public void setOptions() {
        if (radarScreen == null) {
            //Use default settings
            trajectorySel = TerminalControl.trajectorySel;
            pastTrajTime = TerminalControl.pastTrajTime;
            showMva = TerminalControl.showMva;
            showIlsDash = TerminalControl.showIlsDash;
            showUncontrolled = TerminalControl.showUncontrolled;
            rangeCircleDist = TerminalControl.rangeCircleDist;
            colourStyle = TerminalControl.colourStyle;
        } else {
            //Use game settings
            trajectorySel = radarScreen.trajectoryLine;
            pastTrajTime = radarScreen.pastTrajTime;
            showMva = radarScreen.showMva;
            showIlsDash = radarScreen.showIlsDash;
            showUncontrolled = radarScreen.showUncontrolled;
            rangeCircleDist = radarScreen.rangeCircleDist;
            colourStyle = radarScreen.colourStyle;
        }

        trajectoryLine.setSelected(trajectorySel == 0 ? "Off" : trajectorySel + " sec");
        if (pastTrajTime == -1) {
            pastTrajectory.setSelected("All");
        } else if (pastTrajTime == 0) {
            pastTrajectory.setSelected("Off");
        } else {
            pastTrajectory.setSelected(pastTrajTime + " sec");
        }

        mva.setSelected(showMva ? "Show" : "Hide");
        ilsDash.setSelected(showIlsDash ? "Realistic" : "Simple");
        showUncontrolledTrail.setSelected(showUncontrolled ? "Always" : "When selected");
        rangeCircle.setSelected(rangeCircleDist == 0 ? "Off" : rangeCircleDist + "nm");
        colour.setSelectedIndex(colourStyle);
    }

    /** Confirms and applies the changes set */
    @Override
    public void sendChanges() {
        if (radarScreen != null) {
            radarScreen.trajectoryLine = trajectorySel;
            radarScreen.pastTrajTime = pastTrajTime;
            radarScreen.showMva = showMva;
            radarScreen.showIlsDash = showIlsDash;
            radarScreen.showUncontrolled = showUncontrolled;
            boolean rangeDistChanged = radarScreen.rangeCircleDist != rangeCircleDist;
            radarScreen.rangeCircleDist = rangeCircleDist;
            boolean colourChanged = radarScreen.colourStyle != colourStyle;
            radarScreen.colourStyle = colourStyle;

            if (rangeDistChanged) radarScreen.loadRange(); //Reload the range circles in case of any changes
            if (colourChanged) radarScreen.updateColourStyle(); //Update the label colours
        } else {
            TerminalControl.trajectorySel = trajectorySel;
            TerminalControl.pastTrajTime = pastTrajTime;
            TerminalControl.showMva = showMva;
            TerminalControl.showIlsDash = showIlsDash;
            TerminalControl.showUncontrolled = showUncontrolled;
            TerminalControl.rangeCircleDist = rangeCircleDist;
            TerminalControl.colourStyle = colourStyle;

            GameSaver.saveSettings();
        }
    }
}

File: core/src/com/bombbird/terminalcontrol/screens/settingsscreen/categories/DisplaySettingsScreen.kt
package com.bombbird.terminalcontrol.screens.settingsscreen.categories

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.scenes.scene2d.Actor
import com.badlogic.gdx.scenes.scene2d.ui.Image
import com.badlogic.gdx.scenes.scene2d.ui.Label
import com.badlogic.gdx.scenes.scene2d.ui.SelectBox
import com.badlogic.gdx.scenes.scene2d.utils.ChangeListener
import com.bombbird.terminalcontrol.TerminalControl
import com.bombbird.terminalcontrol.screens.gamescreen.RadarScreen
import com.bombbird.terminalcontrol.screens.settingsscreen.SettingsTab
import com.bombbird.terminalcontrol.screens.settingsscreen.SettingsTemplateScreen
import com.bombbird.terminalcontrol.utilities.saving.GameSaver

class DisplaySettingsScreen(game: TerminalControl, radarScreen: RadarScreen?, background: Image?) : SettingsTemplateScreen(game, radarScreen, background) {
    lateinit var trajectoryLine: SelectBox<String>
    lateinit var pastTrajectory: SelectBox<String>
    lateinit var mva: SelectBox<String>
    lateinit var ilsDash: SelectBox<String>
    lateinit var showUncontrolledTrail: SelectBox<String>
    lateinit var rangeCircle: SelectBox<String>
    lateinit var colour: SelectBox<String>
    lateinit var metar: SelectBox<String>
    lateinit var trajectoryLabel: Label
    var trajectorySel = 0
    private lateinit var pastTrajLabel: Label
    var pastTrajTime = 0
    private lateinit var mvaLabel: Label
    var showMva = false
    private lateinit var ilsDashLabel: Label
    var showIlsDash = false
    private lateinit var showUncontrolledLabel: Label
    var showUncontrolled = false
    private lateinit var rangeCircleLabel: Label
    var rangeCircleDist = 0
    private lateinit var colourLabel: Label
    var colourStyle = 0
    private lateinit var metarLabel: Label
    var realisticMetar = false

    init {
        infoString = "Set the game display settings below."
        loadUI(-1200, -200)
        setOptions()
    }

    /** Loads selectBox for display settings  */
    override fun loadBoxes() {
        trajectoryLine = createStandardSelectBox()
        trajectoryLine.setItems("Off", "60 sec", "90 sec", "120 sec", "150 sec")
        trajectoryLine.addListener(object : ChangeListener() {
            override fun changed(event: ChangeEvent, actor: Actor) {
                trajectorySel = if ("Off" == trajectoryLine.selected) {
                    0
                } else {
                    trajectoryLine.selected.split(" ".toRegex()).toTypedArray()[0].toInt()
                }
            }
        })

        pastTrajectory = createStandardSelectBox()
        pastTrajectory.setItems("Off", "60 sec", "120 sec", "All")
        pastTrajectory.addListener(object : ChangeListener() {
            override fun changed(event: ChangeEvent, actor: Actor) {
                pastTrajTime = when (pastTrajectory.selected) {
                    "Off" -> 0
                    "All" -> -1
                    else -> pastTrajectory.selected.split(" ".toRegex()).toTypedArray()[0].toInt()
                }
            }
        })

        mva = createStandardSelectBox()
        mva.setItems("Show", "Hide")
        mva.addListener(object : ChangeListener() {
            override fun changed(event: ChangeEvent, actor: Actor) {
                when (mva.selected) {
                    "Show" -> showMva = true
                    "Hide" -> showMva = false
                    else -> Gdx.app.log(className, "Unknown MVA setting " + mva.selected)
                }
            }
        })

        ilsDash = createStandardSelectBox()
        ilsDash.setItems("Simple", "Realistic")
        ilsDash.addListener(object : ChangeListener() {
            override fun changed(event: ChangeEvent, actor: Actor) {
                when (ilsDash.selected) {
                    "Simple" -> showIlsDash = false
                    "Realistic" -> showIlsDash = true
                    else -> Gdx.app.log(className, "Unknown ILS dash setting " + ilsDash.selected)
                }
            }
        })

        showUncontrolledTrail = createStandardSelectBox()
        showUncontrolledTrail.setItems("When selected", "Always")
        showUncontrolledTrail.addListener(object : ChangeListener() {
            override fun changed(event: ChangeEvent, actor: Actor) {
                when (showUncontrolledTrail.selected) {
                    "Always" -> showUncontrolled = true
                    "When selected" -> showUncontrolled = false
                    else -> Gdx.app.log(className, "Unknown uncontrolled trail setting " + ilsDash.selected)
                }
            }
        })

        rangeCircle = createStandardSelectBox()
        rangeCircle.setItems("Off", "5nm", "10nm", "15nm", "20nm")
        rangeCircle.addListener(object : ChangeListener() {
            override fun changed(event: ChangeEvent, actor: Actor) {
                rangeCircleDist = if ("Off" == rangeCircle.selected) {
                    0
                } else {
                    rangeCircle.selected.replace("nm", "").toInt()
                }
            }
        })

        colour = createStandardSelectBox()
        colour.setItems("More colourful", "More standardised")
        colour.addListener(object : ChangeListener() {
            override fun changed(event: ChangeEvent, actor: Actor) {
                when (colour.selected) {
                    "More colourful" -> colourStyle = 0
                    "More standardised" -> colourStyle = 1
                    else -> Gdx.app.log(className, "Unknown colour style setting " + colour.selected)
                }
            }
        })

        metar = createStandardSelectBox()
        metar.setItems("Simple", "Realistic")
        metar.addListener(object : ChangeListener() {
            override fun changed(event: ChangeEvent, actor: Actor) {
                realisticMetar = "Realistic" == metar.selected
            }
        })
    }

    /** Loads labels for display settings  */
    override fun loadLabel() {
        super.loadLabel()
        trajectoryLabel = Label("Trajectory line: ", labelStyle)
        pastTrajLabel = Label("Aircraft trail: ", labelStyle)
        mvaLabel = Label("MVA altitude: ", labelStyle)
        ilsDashLabel = Label("ILS display: ", labelStyle)
        showUncontrolledLabel = Label("Show uncontrolled:\naircraft trail", labelStyle)
        rangeCircleLabel = Label("Range rings:", labelStyle)
        colourLabel = Label("Colour style:", labelStyle)
        metarLabel = Label("Metar display:", labelStyle)
    }

    /** Loads actors for display settings into tabs  */
    override fun loadTabs() {
        val tab1 = SettingsTab(this, 2)
        tab1.addActors(trajectoryLine, trajectoryLabel)
        tab1.addActors(pastTrajectory, pastTrajLabel)
        tab1.addActors(mva, mvaLabel)
        tab1.addActors(ilsDash, ilsDashLabel)
        tab1.addActors(showUncontrolledTrail, showUncontrolledLabel)
        tab1.addActors(rangeCircle, rangeCircleLabel)
        tab1.addActors(colour, colourLabel)
        tab1.addActors(metar, metarLabel)
        settingsTabs.add(tab1)
    }

    /** Sets relevant options into select boxes  */
    override fun setOptions() {
        if (radarScreen == null) {
            //Use default settings
            trajectorySel = TerminalControl.trajectorySel
            pastTrajTime = TerminalControl.pastTrajTime
            showMva = TerminalControl.showMva
            showIlsDash = TerminalControl.showIlsDash
            showUncontrolled = TerminalControl.showUncontrolled
            rangeCircleDist = TerminalControl.rangeCircleDist
            colourStyle = TerminalControl.colourStyle
            realisticMetar = TerminalControl.realisticMetar
        } else {
            //Use game settings
            trajectorySel = radarScreen.trajectoryLine
            pastTrajTime = radarScreen.pastTrajTime
            showMva = radarScreen.showMva
            showIlsDash = radarScreen.showIlsDash
            showUncontrolled = radarScreen.showUncontrolled
            rangeCircleDist = radarScreen.rangeCircleDist
            colourStyle = radarScreen.colourStyle
            realisticMetar = radarScreen.realisticMetar
        }
        trajectoryLine.selected = if (trajectorySel == 0) "Off" else "$trajectorySel sec"
        when (pastTrajTime) {
            -1 -> pastTrajectory.setSelected("All")
            0 -> pastTrajectory.setSelected("Off")
            else -> pastTrajectory.setSelected("$pastTrajTime sec")
        }
        mva.selected = if (showMva) "Show" else "Hide"
        ilsDash.selected = if (showIlsDash) "Realistic" else "Simple"
        showUncontrolledTrail.selected = if (showUncontrolled) "Always" else "When selected"
        rangeCircle.selected = if (rangeCircleDist == 0) "Off" else rangeCircleDist.toString() + "nm"
        colour.selectedIndex = colourStyle
        metar.selected = if (realisticMetar) "Realistic" else "Simple"
    }

    /** Confirms and applies the changes set  */
    override fun sendChanges() {
        if (radarScreen != null) {
            radarScreen.trajectoryLine = trajectorySel
            radarScreen.pastTrajTime = pastTrajTime
            radarScreen.showMva = showMva
            radarScreen.showIlsDash = showIlsDash
            radarScreen.showUncontrolled = showUncontrolled
            val rangeDistChanged = radarScreen.rangeCircleDist != rangeCircleDist
            radarScreen.rangeCircleDist = rangeCircleDist
            val colourChanged = radarScreen.colourStyle != colourStyle
            radarScreen.colourStyle = colourStyle
            val metarChanged = radarScreen.realisticMetar != realisticMetar
            radarScreen.realisticMetar = realisticMetar
            Gdx.app.postRunnable {
                if (rangeDistChanged) radarScreen.loadRange() //Reload the range circles in case of any changes
                if (colourChanged) radarScreen.updateColourStyle() //Update the label colours
                if (metarChanged) radarScreen.ui.updateMetar() //Update metar display
            }
        } else {
            TerminalControl.trajectorySel = trajectorySel
            TerminalControl.pastTrajTime = pastTrajTime
            TerminalControl.showMva = showMva
            TerminalControl.showIlsDash = showIlsDash
            TerminalControl.showUncontrolled = showUncontrolled
            TerminalControl.rangeCircleDist = rangeCircleDist
            TerminalControl.colourStyle = colourStyle
            TerminalControl.realisticMetar = realisticMetar
            GameSaver.saveSettings()
        }
    }
}